cardano-crypto-class/src/Cardano/Crypto/Libsodium/Hash.hs

```diff
@@ -0,0 +1,134 @@
+{-# LANGUAGE CApiFFI #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE InstanceSigs #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE RoleAnnotations #-}
+{-# LANGUAGE TypeApplications #-}
+module Cardano.Crypto.Libsodium.Hash (
+    SodiumHashAlgorithm (..),
+    digestMLockedStorable,
+    digestMLockedFB,
+    digestMLockedBS,
+    expandHash,
+) where
+
+import Control.Exception (bracket)
+import Control.Monad (unless)
+import Data.Proxy (Proxy (..))
+import Foreign.C.Error (errnoToIOError, getErrno)
+import Foreign.C.Types (CSize)
+import Foreign.Ptr (Ptr, castPtr, nullPtr, plusPtr)
+import Foreign.Storable (Storable (sizeOf, poke))
+import Data.Word (Word8)
+import GHC.IO.Exception (ioException)
+import GHC.TypeLits
+import System.IO.Unsafe (unsafeDupablePerformIO)
+import GHC.IO.Handle.Text (memcpy)
+
+import qualified Data.ByteString as SB
+
+import Cardano.Crypto.Hash (HashAlgorithm, SHA256, Blake2b_256)
+import Cardano.Crypto.FiniteBytes (FiniteBytes)
+import Cardano.Crypto.Libsodium.C
+import Cardano.Crypto.Libsodium.Memory.Internal
+import Cardano.Crypto.Libsodium.MLockedBytes.Internal
+
+-------------------------------------------------------------------------------
+-- Type-Class
+-------------------------------------------------------------------------------
+
+class (HashAlgorithm h, KnownNat (SizeHash h)) => SodiumHashAlgorithm h where
+    -- | The size in bytes of the output of 'digest'
+    type SizeHash h :: Nat
+
+    digestMLocked
+        :: proxy h
+        -> Ptr a  -- ^ input
+        -> Int    -- ^ input length
+        -> IO (MLockedFiniteBytes (SizeHash h))
+
+    -- TODO: provide interface for multi-part?
+    -- That will be useful to hashing ('1' <> oldseed).
+
+digestMLockedStorable
+    :: forall h a proxy. (SodiumHashAlgorithm h, Storable a)
+    => proxy h -> Ptr a -> IO (MLockedFiniteBytes (SizeHash h))
+digestMLockedStorable p ptr =
+  digestMLocked p ptr ((sizeOf (undefined :: a)))
+
+digestMLockedFB
+    :: forall h n proxy. (SodiumHashAlgorithm h, KnownNat n)
+    => proxy h -> Ptr (FiniteBytes n) -> IO (MLockedFiniteBytes (SizeHash h))
+digestMLockedFB = digestMLockedStorable
+
+digestMLockedBS
+    :: forall h proxy. (SodiumHashAlgorithm h)
+    => proxy h -> SB.ByteString -> IO (MLockedFiniteBytes (SizeHash h))
+digestMLockedBS p bs = SB.useAsCStringLen bs $ \(ptr, len) -> do
+    digestMLocked p (castPtr ptr) len
+
+-------------------------------------------------------------------------------
+-- Hash expansion
```

Comment:
Can you provide some more detail here? What is hash expansion, and why would I want to do it?

Reply:
Oh, or this is specific to the SUM KES construction?

Reply:
Yes this is specific to  SUM KES. The current name for this is `splitSeed`, but that one isn't good either as it uses *some* hashing (and produces seed of size specific to the hash function).

Better name would help.

FWIW, sodium's blake2b implementation allows to pass *key* to the hash function, thus AFAICS salting the result, which we do here manually. See https://libsodium.gitbook.io/doc/hashing/generic_hashing#usage